In [108]:
import pandas as pd

In [109]:
df = pd.read_csv("../data/raw/WA_Fn-UseC_-Telco-Customer-Churn.csv")

In [110]:
df.shape

(7043, 21)

In [111]:
df.columns

Index(['customerID', 'gender', 'SeniorCitizen', 'Partner', 'Dependents',
       'tenure', 'PhoneService', 'MultipleLines', 'InternetService',
       'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport',
       'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling',
       'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn'],
      dtype='str')

In [112]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   str    
 1   gender            7043 non-null   str    
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   str    
 4   Dependents        7043 non-null   str    
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   str    
 7   MultipleLines     7043 non-null   str    
 8   InternetService   7043 non-null   str    
 9   OnlineSecurity    7043 non-null   str    
 10  OnlineBackup      7043 non-null   str    
 11  DeviceProtection  7043 non-null   str    
 12  TechSupport       7043 non-null   str    
 13  StreamingTV       7043 non-null   str    
 14  StreamingMovies   7043 non-null   str    
 15  Contract          7043 non-null   str    
 16  PaperlessBilling  7043 non-null   str    
 17  Paymen

In [113]:
df.head()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


In [114]:
df["Churn"].value_counts()

Churn
No     5174
Yes    1869
Name: count, dtype: int64

In [115]:
df["Churn"].value_counts(normalize=True) * 100

Churn
No     73.463013
Yes    26.536987
Name: proportion, dtype: float64

In [116]:
df["Contract"].value_counts()

Contract
Month-to-month    3875
Two year          1695
One year          1473
Name: count, dtype: int64

In [117]:
pd.crosstab(df["Contract"], df["Churn"])

Churn,No,Yes
Contract,,
Month-to-month,2220,1655
One year,1307,166
Two year,1647,48


In [118]:
pd.crosstab(
    df["Contract"],
    df["Churn"],
    normalize="index"
) * 100

Churn,No,Yes
Contract,,
Month-to-month,57.290323,42.709677
One year,88.730482,11.269518
Two year,97.168142,2.831858


## Análise de Churn por Tipo de Contrato

A análise inicial mostra uma diferença significativa na taxa de churn entre os tipos de contrato.

Clientes com contrato Month-to-month apresentam taxa de churn de 42,71%, enquanto clientes com contratos de One year e Two year apresentam taxas de 11,27% e 2,83%, respectivamente.

Esse resultado indica uma forte associação entre tipo de contrato e cancelamento de clientes, que será investigada em conjunto com outras características da base.

In [119]:
df["tenure"].describe()

count    7043.000000
mean       32.371149
std        24.559481
min         0.000000
25%         9.000000
50%        29.000000
75%        55.000000
max        72.000000
Name: tenure, dtype: float64

In [120]:
df["tenure"].value_counts().sort_index()

tenure
0      11
1     613
2     238
3     200
4     176
     ... 
68    100
69     95
70    119
71    170
72    362
Name: count, Length: 73, dtype: int64

In [121]:
pd.crosstab(df["tenure"],df["Churn"], normalize="index") * 100

Churn,No,Yes
tenure,,
0,100.000000,0.000000
1,38.009788,61.990212
2,48.319328,51.680672
3,53.000000,47.000000
4,52.840909,47.159091
...,...,...
68,91.000000,9.000000
69,91.578947,8.421053
70,90.756303,9.243697


In [122]:
df["TenureGroup"] = pd.cut(
    df["tenure"],
    bins=[-1, 12, 24, 48, 72],
    labels=["0-12 meses", "13-24 meses", "25-48 meses", "49-72 meses"]
)

In [123]:
# Taxa de churn dentro de cada combinação de faixa de relacionamento e tipo de contrato

pd.crosstab( 
    df["TenureGroup"],
    df["Churn"],
    normalize="index"
) * 100

Churn,No,Yes
TenureGroup,,
0-12 meses,52.561757,47.438243
13-24 meses,71.289062,28.710938
25-48 meses,79.611041,20.388959
49-72 meses,90.486824,9.513176


In [124]:
df["TenureGroup"].value_counts().sort_index()

TenureGroup
0-12 meses     2186
13-24 meses    1024
25-48 meses    1594
49-72 meses    2239
Name: count, dtype: int64

## Análise de Churn por Tempo de Relacionamento

A taxa de churn apresenta uma relação inversa com o tempo de relacionamento. Clientes com 0–12 meses de relacionamento apresentam churn de 47,44%, enquanto clientes com 49–72 meses apresentam 9,53%.

Como o grupo de 0–12 meses possui 2.186 clientes, esse resultado não está concentrado em uma amostra pequena. O início do relacionamento representa, portanto, um importante ponto de atenção para estratégias de retenção.

Na próxima etapa, será investigada a relação entre tempo de relacionamento e tipo de contrato.

Base de clientes
      ↓
26,54% de churn geral
      ↓
Month-to-month → 42,71%
      ↓
0–12 meses → 47,44%
      ↓
Pergunta:
"Clientes novos + contrato mensal possuem risco ainda maior?"

In [125]:
# dentro de cada faixa de tenure e tipo de contrato, existe diferença na taxa de churn?

pd.crosstab(
    [df["TenureGroup"], df["Contract"]],
     df["Churn"],
     normalize="index"
) * 100

Churn                               No        Yes
TenureGroup Contract                             
0-12 meses  Month-to-month   48.645938  51.354062
            One year         89.516129  10.483871
            Two year        100.000000   0.000000
13-24 meses Month-to-month   62.279512  37.720488
            One year         91.878173   8.121827
            Two year        100.000000   0.000000
25-48 meses Month-to-month   67.082294  32.917706
            One year         89.382239  10.617761
            Two year         97.810219   2.189781
49-72 meses Month-to-month   73.976608  26.023392
            One year         87.066246  12.933754
            Two year         96.674584   3.325416

In [126]:
# Volume de clientes de cada combinação

pd.crosstab(
    [df["TenureGroup"], df["Contract"]],
    df["Churn"]
)

Churn                         No   Yes
TenureGroup Contract                  
0-12 meses  Month-to-month   970  1024
            One year         111    13
            Two year          68     0
13-24 meses Month-to-month   459   278
            One year         181    16
            Two year          90     0
25-48 meses Month-to-month   538   264
            One year         463    55
            Two year         268     6
49-72 meses Month-to-month   253    89
            One year         552    82
            Two year        1221    42

## Relação entre Tempo de Relacionamento e Tipo de Contrato

O cruzamento entre tempo de relacionamento, tipo de contrato e churn reforça os padrões observados anteriormente.

Clientes com contrato Month-to-month apresentam as maiores taxas de churn em todas as faixas de relacionamento analisadas. O grupo de maior atenção é formado por clientes com 0–12 meses de relacionamento e contrato Month-to-month, que apresenta taxa de churn de 51,35% e reúne 1.994 clientes.

Os resultados indicam uma associação relevante entre menor tempo de relacionamento, contratos mensais e maior ocorrência de churn. Entretanto, esses fatores não devem ser interpretados isoladamente nem como evidência de causalidade.

In [127]:
# A taxa de churn é diferente entre clientes de mensalidade baixa, média e alta?

df["MonthlyChargesGroup"] = pd.cut(
    df["MonthlyCharges"],
    bins=[0, 35.50, 89.85, float("inf")],
    labels=["Baixa", "Média", "Alta"]
)

In [128]:
df["MonthlyChargesGroup"].value_counts().sort_index()

MonthlyChargesGroup
Baixa    1762
Média    3523
Alta     1758
Name: count, dtype: int64

In [129]:
# A mensalidade mais alta está associada a maior churn?

pd.crosstab(
    df["MonthlyChargesGroup"],
    df["Churn"],
    normalize="index"
) * 100

Churn,No,Yes
MonthlyChargesGroup,,
Baixa,88.762770,11.237230
Média,68.975305,31.024695
Alta,67.121729,32.878271


In [130]:
# Quantos clientes existem em cada grupo e quantos efetivamente cancelaram.

pd.crosstab(
    df["MonthlyChargesGroup"],
    df["Churn"]
)

Churn,No,Yes
MonthlyChargesGroup,,
Baixa,1564,198
Média,2430,1093
Alta,1180,578


In [131]:
# MonthlyCharges × Contract

pd.crosstab(
    df["Contract"],
    df["MonthlyChargesGroup"],
    normalize="index"
) * 100

MonthlyChargesGroup,Baixa,Média,Alta
Contract,,,
Month-to-month,18.658065,58.890323,22.451613
One year,26.748133,42.973523,30.278344
Two year,38.053097,35.870206,26.076696


In [132]:
# O tipo de serviço de internet está associado a diferentes taxas de churn?

pd.crosstab(
    df["InternetService"],
    df["Churn"],
    normalize="index"
) * 100

Churn,No,Yes
InternetService,,
DSL,81.040892,18.959108
Fiber optic,58.107235,41.892765
No,92.595020,7.404980


## Análise de Churn por Tipo de Serviço de Internet

O tipo de serviço de internet apresenta diferenças relevantes na taxa de churn.

Clientes com serviço Fiber optic apresentam churn de 41,89%, enquanto clientes DSL apresentam 18,59%. Clientes que não possuem serviço de internet apresentam a menor taxa observada, de 7,40%.

O resultado indica uma associação relevante entre o tipo de serviço de internet e o churn. Entretanto, não é possível afirmar causalidade apenas com essa análise, sendo necessário investigar a relação dessa variável com contrato, tempo de relacionamento e mensalidade.

In [133]:
#Clientes sem serviço de segurança online apresentam maior churn?

pd.crosstab(
    df["OnlineSecurity"],
    df["Churn"],
    normalize="index"
) * 100

Churn,No,Yes
OnlineSecurity,,
No,58.233276,41.766724
No internet service,92.595020,7.404980
Yes,85.388806,14.611194


## Análise de Churn por Serviço de Segurança Online

Clientes sem o serviço OnlineSecurity apresentam taxa de churn de 41,77%, enquanto clientes que possuem o serviço apresentam taxa de 14,61%.

O resultado indica uma associação relevante entre a ausência do serviço de segurança online e maior churn. Entretanto, essa relação pode estar associada a outras características dos clientes, como tipo de serviço de internet, contrato e mensalidade, que também apresentaram diferenças nas taxas de churn.

In [134]:
# Clientes sem TechSupport apresentam maior taxa de churn?

pd.crosstab(
    df["TechSupport"],
    df["Churn"],
    normalize="index"
) * 100

Churn,No,Yes
TechSupport,,
No,58.364526,41.635474
No internet service,92.595020,7.404980
Yes,84.833659,15.166341


#                     MAIOR CHURN
                         │
          ┌──────────────┼──────────────┐
          ↓              ↓              ↓
   Month-to-month   0–12 meses     Fiber optic
      42,71%          47,44%          41,89%
          │
          ├── OnlineSecurity = No → 41,77%
          │
          └── TechSupport = No → 41,64%

## Análise de Churn por Suporte Técnico

Clientes sem o serviço TechSupport apresentam taxa de churn de 41,64%, enquanto clientes que possuem o serviço apresentam taxa de 15,17%.

O resultado é semelhante ao observado para OnlineSecurity, reforçando a existência de uma associação entre a ausência de serviços adicionais e maior churn.

Entretanto, essas variáveis podem estar relacionadas entre si e com outras características dos clientes, como tipo de internet, contrato e mensalidade. Portanto, os resultados não devem ser interpretados como relações causais isoladas.

In [135]:
# Cliente com contrato Month-to-month + até 12 meses de relacionamento + sem OnlineSecurity + sem TechSupport. (hipótese analítica)

perfil_critico = (
    (df["Contract"] == "Month-to-month") &
    (df["TenureGroup"] == "0-12 meses") &
    (df["OnlineSecurity"] == "No") &
    (df["TechSupport"] == "No")
)

df["PerfilRisco"] = perfil_critico.map({
    True: "Perfil crítico",
    False: "Demais clientes"
})

In [136]:
df["PerfilRisco"].value_counts()

PerfilRisco
Demais clientes    5855
Perfil crítico     1188
Name: count, dtype: int64

In [137]:
# Qual é a taxa de churn dos clientes que possuem esse perfil em comparação aos demais?

pd.crosstab(
    df["PerfilRisco"],
    df["Churn"],
    normalize="index"
) * 100

Churn,No,Yes
PerfilRisco,,
Demais clientes,81.161401,18.838599
Perfil crítico,35.521886,64.478114


In [138]:
# Quantos dos 1.188 clientes desse perfil realmente cancelaram?

pd.crosstab(
    df["PerfilRisco"],
    df["Churn"]
)

Churn,No,Yes
PerfilRisco,,
Demais clientes,4752,1103
Perfil crítico,422,766


## Identificação de Perfil Crítico de Churn

A combinação de contrato Month-to-month, relacionamento de 0–12 meses, ausência de OnlineSecurity e ausência de TechSupport foi utilizada para identificar um perfil exploratório de maior risco.

O grupo identificado possui 1.188 clientes, dos quais 766 cancelaram, resultando em uma taxa de churn de 64,48%.

Entre os demais 5.855 clientes, a taxa de churn foi de 18,84%.

Embora esse perfil represente aproximadamente 16,9% da base, ele concentra cerca de 41% dos cancelamentos observados. O resultado sugere um segmento prioritário para investigação e possíveis ações de retenção, sem estabelecer relação causal ou capacidade preditiva.

In [139]:
# A taxa de churn varia de acordo com a forma de pagamento utilizada pelo cliente?

pd.crosstab(
    df["PaymentMethod"],
    df["Churn"],
    normalize="index"
) * 100

Churn,No,Yes
PaymentMethod,,
Bank transfer (automatic),83.290155,16.709845
Credit card (automatic),84.756899,15.243101
Electronic check,54.714588,45.285412
Mailed check,80.893300,19.106700


In [140]:
# Quantos clientes utilizam cada forma de pagamento e quantos cancelaram?

pd.crosstab(
    df["PaymentMethod"],
    df["Churn"]
)


Churn,No,Yes
PaymentMethod,,
Bank transfer (automatic),1286,258
Credit card (automatic),1290,232
Electronic check,1294,1071
Mailed check,1304,308


## Análise de Churn por Forma de Pagamento

A forma de pagamento apresenta diferenças relevantes na taxa de churn.

Clientes que utilizam Electronic check apresentam churn de 45,29%, significativamente superior às demais formas de pagamento, que apresentam taxas entre 15,24% e 19,11%.

O grupo Electronic check possui 2.365 clientes, dos quais 1.071 cancelaram. Esses cancelamentos representam aproximadamente 57% de todos os cancelamentos observados na base.

O resultado indica uma forte associação entre o uso de Electronic check e churn, tornando esse grupo relevante para investigação de possíveis fatores comportamentais e operacionais relacionados à retenção. Entretanto, não é possível estabelecer causalidade apenas com essa análise.

In [141]:
# Forma de pagamento × Contrato

pd.crosstab(
    df["Contract"],
    df["PaymentMethod"],
    normalize="index"
) * 100


PaymentMethod,Bank transfer (automatic),Credit card (automatic),Electronic check,Mailed check
Contract,,,,
Month-to-month,15.200000,14.012903,47.741935,23.045161
One year,26.544467,27.019688,23.557366,22.878479
Two year,33.274336,34.277286,9.911504,22.536873


In [142]:
# Churn dentro de cada contrato

pd.crosstab(
    [df["Contract"], df["PaymentMethod"]],
    df["Churn"],
    normalize="index"
) * 100

Churn                                            No        Yes
Contract       PaymentMethod                                  
Month-to-month Bank transfer (automatic)  65.874363  34.125637
               Credit card (automatic)    67.219153  32.780847
               Electronic check           46.270270  53.729730
               Mailed check               68.421053  31.578947
One year       Bank transfer (automatic)  90.281330   9.718670
               Credit card (automatic)    89.698492  10.301508
               Electronic check           81.556196  18.443804
               Mailed check               93.175074   6.824926
Two year       Bank transfer (automatic)  96.631206   3.368794
               Credit card (automatic)    97.762478   2.237522
               Electronic check           92.261905   7.738095
               Mailed check               99.214660   0.785340

In [143]:
# Volume de clientes em cada combinação.

pd.crosstab(
    df["Contract"],
    df["PaymentMethod"]
)

PaymentMethod,Bank transfer (automatic),Credit card (automatic),Electronic check,Mailed check
Contract,,,,
Month-to-month,589,543,1850,893
One year,391,398,347,337
Two year,564,581,168,382


In [144]:
# Fatores do perfil crítico com Electronic check

pd.crosstab(
    df["PerfilRisco"],
    df["PaymentMethod"]
)

PaymentMethod,Bank transfer (automatic),Credit card (automatic),Electronic check,Mailed check
PerfilRisco,,,,
Demais clientes,1436,1421,1631,1367
Perfil crítico,108,101,734,245


In [145]:
# Perfil crítico + Electronic check

perfil_critico_echeck = (
    (df["PerfilRisco"] == "Perfil crítico") &
    (df["PaymentMethod"] == "Electronic check")
)

pd.crosstab(
    perfil_critico_echeck,
    df["Churn"],
    normalize="index"
) * 100

Churn,No,Yes
row_0,,
False,78.491045,21.508955
True,30.245232,69.754768


## Segmento de Maior Risco Identificado

A análise combinou o perfil crítico previamente identificado com a forma de pagamento Electronic check.

O grupo formado por clientes com contrato Month-to-month, relacionamento de 0–12 meses, ausência de OnlineSecurity e ausência de TechSupport apresentou 69,75% de churn quando também utilizava Electronic check.

Esse segmento possui 734 clientes, dos quais aproximadamente 512 cancelaram. Os cancelamentos desse grupo representam cerca de 27,4% de todos os cancelamentos observados na base.

O resultado identifica um segmento de alta concentração de churn e pode orientar futuras análises e estratégias de retenção. Entretanto, trata-se de uma segmentação exploratória e não de um modelo preditivo ou evidência de causalidade.

# Principais Indicadores de Churn

Nesta etapa são consolidados os principais indicadores identificados durante a análise exploratória.

Os indicadores serão utilizados posteriormente para comparação no Power BI e para apoiar as recomendações de negócio.

In [146]:
# KPIs

total_clientes = len(df)

total_churn = (df["Churn"] == "Yes").sum()

taxa_churn = (total_churn / total_clientes) * 100

clientes_perfil_critico = (df["PerfilRisco"] == "Perfil crítico").sum()

churn_perfil_critico = (
    df.loc[df["PerfilRisco"] == "Perfil crítico", "Churn"]
    .eq("Yes")
    .mean()
) * 100

clientes_perfil_echeck = perfil_critico_echeck.sum()

churn_perfil_echeck = (
    df.loc[perfil_critico_echeck, "Churn"]
    .eq("Yes")
    .mean()
) * 100

print(f"Total de clientes: {total_clientes:,}")
print(f"Total de cancelamentos: {total_churn:,}")
print(f"Taxa de churn: {taxa_churn:.2f}%")
print(f"Clientes no perfil crítico: {clientes_perfil_critico:,}")
print(f"Churn do perfil crítico: {churn_perfil_critico:.2f}%")
print(f"Clientes no perfil crítico + Electronic check: {clientes_perfil_echeck:,}")
print(f"Churn do perfil crítico + Electronic check: {churn_perfil_echeck:.2f}%")

Total de clientes: 7,043
Total de cancelamentos: 1,869
Taxa de churn: 26.54%
Clientes no perfil crítico: 1,188
Churn do perfil crítico: 64.48%
Clientes no perfil crítico + Electronic check: 734
Churn do perfil crítico + Electronic check: 69.75%


# Principais Insights da Análise Exploratória

A análise exploratória identificou diferenças relevantes nas taxas de churn entre diferentes perfis de clientes.

## Principais achados

1. A taxa geral de churn da base é de 26,54%.

2. Clientes com contrato Month-to-month apresentam churn de 42,71%, significativamente superior aos clientes com contratos de maior duração.

3. Clientes com 0–12 meses de relacionamento apresentam churn de 47,44%, indicando maior concentração de cancelamentos no início do relacionamento.

4. Clientes com serviço Fiber optic apresentam churn de 41,89%, enquanto clientes DSL apresentam 18,59%.

5. Clientes sem OnlineSecurity apresentam churn de 41,77%, contra 14,61% entre aqueles que possuem o serviço.

6. Clientes sem TechSupport apresentam churn de 41,64%, contra 15,17% entre aqueles que possuem o serviço.

7. Clientes que utilizam Electronic check apresentam churn de 45,29%, superior às demais formas de pagamento.

8. A combinação de contrato Month-to-month, relacionamento de 0–12 meses, ausência de OnlineSecurity e ausência de TechSupport identificou um perfil exploratório com churn de 64,48%.

9. Quando esse perfil também utiliza Electronic check, a taxa de churn observada aumenta para 69,75%.

## Considerações analíticas

Os resultados indicam associações relevantes entre características dos clientes e churn. Entretanto, as análises realizadas são exploratórias e não permitem estabelecer relações de causalidade.

Os segmentos identificados devem ser utilizados como hipóteses para investigação e como possíveis grupos prioritários em estratégias de retenção.

# Encerramento do Data Profiling

A etapa de Data Profiling permitiu compreender a estrutura da base, validar sua qualidade inicial e identificar os principais padrões relacionados ao cancelamento de clientes.

Os resultados obtidos serão utilizados como hipóteses nas próximas etapas do projeto.

Na sequência será realizada a preparação e limpeza dos dados, seguida por uma análise exploratória mais estruturada, análise em SQL e construção do dashboard no Power BI.